tests/ops/service/connectors/test_connector_template_loaders.py

```diff
@@ -30,3 +47,176 @@ def test_file_connector_template_loader_connector_not_found(self):
         connector_templates = loader.get_connector_templates()
 
         assert connector_templates.get("not_found") is None
+
+
+class TestCustomConnectorTemplateLoader:
+    def test_file_connector_template_loader_no_templates(self):
+        loader = CustomConnectorTemplateLoader()
+        connector_templates = loader.get_connector_templates()
+        assert connector_templates == {}
+
+    def test_file_connector_template_loader_invalid_template(
+        self, db, custom_dataset, custom_icon, custom_functions
+    ):
+        # save custom connector template to the database
+        custom_template = CustomConnectorTemplate(
+            key="custom",
+            name="Custom",
+            config="custom_config",
+            dataset=custom_dataset,
+            icon=custom_icon,
+            functions=custom_functions,
+        )
+        custom_template.save(db=db)
+
+        # verify the custom functions aren't loaded if the template is invalid
+        loader = CustomConnectorTemplateLoader()
+        connector_templates = loader.get_connector_templates()
+        assert connector_templates == {}
+
+        with pytest.raises(NoSuchSaaSRequestOverrideException):
+            SaaSRequestOverrideFactory.get_override(
+                "custom_user_access", SaaSRequestType.READ
+            )
+
+        # assert the strategy was not registered
+        authentication_strategies = AuthenticationStrategy.get_strategies()
+        assert "custom" not in [strategy.name for strategy in authentication_strategies]
+
+    def test_file_connector_template_loader_invalid_functions(
+        self, db, custom_config, custom_dataset, custom_icon
+    ):
+        # save custom connector template to the database
+        custom_template = CustomConnectorTemplate(
+            key="custom",
+            name="Custom",
+            config=custom_config,
+            dataset=custom_dataset,
+            icon=custom_icon,
+            functions="custom_functions",
+        )
+        custom_template.save(db=db)
+
+        # verify nothing is loaded if the custom functions fail to load
+        loader = CustomConnectorTemplateLoader()
+        connector_templates = loader.get_connector_templates()
+        assert connector_templates == {}
+
+    def test_file_connector_template_loader_custom_connector_functions_disabled(
+        self, db, custom_config, custom_dataset, custom_icon, custom_functions
+    ):
+        CONFIG.security.allow_custom_connector_functions = False
+
+        # save custom connector template to the database
+        custom_template = CustomConnectorTemplate(
+            key="custom",
+            name="Custom",
+            config=custom_config,
+            dataset=custom_dataset,
+            icon=custom_icon,
+            functions=custom_functions,
+        )
+        custom_template.save(db=db)
+
+        # load custom connector templates from the database
+        loader = CustomConnectorTemplateLoader()
```

Comment:
ok, looking a bit closer at the rest of the PR i realize that:

- we do _log_ an error here at load time
- this also isn't the code path that will be invoked for the actual dynamic registration of templates via an endpoint, this is just loading their persisted records from the db the next application bootup

so i retract most of my comment above - i think the UX is OK for this case. that being said, i do have other concerns i raised (in separate comments) about ensuring our dynamic registration path goes thru the same "validations" that we do here in the app bootstrap path. if/when we add the validation in to that path (which i think is a _must_), then in _those_ cases, i do think we need to fail fast and report that to the user immediately. i.e., if a user tries to upload a saas template with a custom function while that feature is disabled via config property, that operation should completely fail and the reason should reported clearly to the user. 

while i know your code here in this PR hasn't really gotten to that point yet, i do think what you've got here will need to be refactored a bit to support that UX, so it may be prudent to address it now.